src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -114,89 +104,94 @@ func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, pa
 
 	// Results is closed by execute
 	results := make(chan executor.Query)
-	// Block series slices are sorted by start time
-	seriesMap := make(map[string][]block.Series)
 	go h.engine.ExecuteExpr(ctx, parser, opts, params, results)
 
+	// Block slices are sorted by start time
+	// TODO: Pooling
+	sortedBlockList := make([]block.Block, 0, 10)
 	for result := range results {
 		if result.Err != nil {
-			return nil, err
+			return nil, result.Err
 		}
 
 		blocks := result.Result.Blocks()
 		// TODO(nikunj): Stream blocks to client
-		for blk := range blocks {
-			iter := blk.SeriesIter()
-			for iter.Next() {
-				insertSeriesInMap(iter.Current(), seriesMap)
+		for blkResult := range blocks {
+			if blkResult.Err != nil {
+				return nil, blkResult.Err
 			}
 
-			// TODO(nikunj): Figure out how to close this outside
-			blk.Close()
+			// Insert blocks sorted by start time
+			sortedBlockList, err = insertSortedBlock(blkResult.Block, sortedBlockList)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 
-	return seriesMapToSeriesList(seriesMap)
+	seriesList, err := sortedBlocksToSeriesList(sortedBlockList)
+	for _, b := range sortedBlockList {
+		b.Close()
```

Comment:
Unfortunately I don't think this will close the blocks if any have errored and we shortcircuit the `return`. Might be an idea to capture any `err` in the loop, then break out and close the blocks in that case too

Alternatively, can do something like `defer results.Close()` and add a Close() method on `results` that would propagate to its blocks?